corehq/util/tests/test_queries.py

```diff
@@ -5,45 +5,82 @@
 
 
 class TestQuerysetToIterator(TestCase):
-    @classmethod
-    def setUpClass(cls):
-        super().setUpClass()
-        cls.users = [
-            User.objects.create_user(f'user{i}@example.com', last_name="Tenenbaum")
-            for i in range(1, 11)
-        ]
 
-    @classmethod
-    def tearDownClass(cls):
-        for user in cls.users:
-            user.delete()
-        super().tearDownClass()
+    def test_correct_results_are_returned(self):
+        query = User.objects.filter(last_name="Tenenbaum")
+
+        results = list(queryset_to_iterator(query, User, limit=10))
+
+        self.assertEqual(
+            [u.username for u in results],
+            [u.username for u in self.users],
+        )
+
+    def test_results_returned_in_one_query_if_limit_is_greater_than_result_size(self):
+        query = User.objects.filter(last_name="Tenenbaum")
+
+        with self.assertNumQueries(1):
+            # query 1: Users 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
+            results = list(queryset_to_iterator(query, User, limit=11))
+
+        self.assertEqual(len(results), 10)
+
+    def test_results_returned_in_two_queries_if_limit_is_equal_to_result_size(self):
+        query = User.objects.filter(last_name="Tenenbaum")
 
-    def test_queryset_to_iterator(self):
+        with self.assertNumQueries(2):
+            # query 1: Users 1, 2, 3, 4, 5, 6, 7, 8, 9, 10
+            # query 2: Check that there are no users past #10
+            results = list(queryset_to_iterator(query, User, limit=10))
+
+        self.assertEqual(len(results), 10)
+
+    def test_results_return_in_three_queries_if_limit_is_less_than_or_equal_to_half_of_result_size(self):
         query = User.objects.filter(last_name="Tenenbaum")
-        self.assertEqual(query.count(), 10)
 
-        with self.assertNumQueries(4):
+        with self.assertNumQueries(3):
```

Comment:
@esoergel given this behavior was defined in tests as well, I wanted to clarify if it was intentional. Basically, is there a concern that breaking the pagination loop when the # of docs returned in for a page is less than the limit set could lead to prematurely exiting pagination? This maps to the addition of:
```
        if doc_count < limit:
            break
```
in `queryset_to_iterator`.

Reply:
(also I assumed you added this based on commit history, but I admittedly did not look at it very hard so if you don't have context, totally fine)

Reply:
Yeah I think I added this - that does seem reasonable, not sure why I didn't do that in the first place.  Reviewing this now, my initial thought was that the dataset can change while the query is executing, which could cause weirdness, but the limit is part of that last query, so the number of results returned should be a valid data point in determining whether there are any remaining.